RecoCTPPS/PixelLocal/src/RPixPlaneCombinatoryTracking.cc

```diff
@@ -0,0 +1,513 @@
+  if(verbosity_>=2) edm::LogInfo("RPixPlaneCombinatoryTracking")<<"Number of combinations = "<<numberOfCombinations;
+  possiblePlaneCombinations_.reserve(numberOfCombinations);
+
+  possiblePlaneCombinations_ = getPlaneCombinations(listOfAllPlanes_,trackMinNumberOfPoints_);
+
+  if(verbosity_>=2) {
+    for( const auto & vec : possiblePlaneCombinations_){
+      for( const auto & num : vec){
+        edm::LogInfo("RPixPlaneCombinatoryTracking")<<num<<" - ";
+      }
+      edm::LogInfo("RPixPlaneCombinatoryTracking");
+    }
+  }
+
+}
+
+//------------------------------------------------------------------------------------------------//
+    
+//This function produces all the possible plane combinations extracting numberToExtract planes over numberOfPlanes planes
+RPixPlaneCombinatoryTracking::PlaneCombinations
+RPixPlaneCombinatoryTracking::getPlaneCombinations(const std::vector<uint32_t> &inputPlaneList, uint32_t numberToExtract)
+{
+    uint32_t numberOfPlanes = inputPlaneList.size();
+    std::string bitmask(numberToExtract, 1); // numberToExtract leading 1's
+    bitmask.resize(numberOfPlanes, 0); // numberOfPlanes-numberToExtract trailing 0's
+    PlaneCombinations planeCombinations;
+
+    // store the combination and permute bitmask
+    do {
+      std::vector<uint32_t> tmpPlaneList;
+      for (uint32_t i = 0; i < numberOfPlanes; ++i) { // [0..numberOfPlanes-1] integers
+        if (bitmask[i]) tmpPlaneList.push_back(inputPlaneList.at(i));
+      }
+        planeCombinations.push_back(tmpPlaneList);
+    } while (std::prev_permutation(bitmask.begin(), bitmask.end()));
+
+    return planeCombinations;
+}
+
+
+//------------------------------------------------------------------------------------------------//
+
+//This function calls it self in order to get all the possible combinations of hits having a certain selected number of planes
+//This function avoids to write for loops in cascade which will not allow to define arbitrarily the minimum number of planes to use
+//The output is stored in a map containing the vector of points and as a key the map of the point forming this vector.
+//This allows to erase the points already used for the track fit
+void RPixPlaneCombinatoryTracking::getHitCombinations(
+  const std::map<CTPPSPixelDetId, PointInPlaneList > &mapOfAllHits, 
+  std::map<CTPPSPixelDetId, PointInPlaneList >::iterator mapIterator,
+  HitReferences tmpHitPlaneMap,
+  const PointInPlaneList &tmpHitVector,
+  PointAndReferenceMap &outputMap)
+{
+    //At this point I selected one hit per plane
+    if (mapIterator == mapOfAllHits.end())
+    {
+        outputMap[tmpHitPlaneMap] = tmpHitVector;
+        return;
+    }
+    for (size_t i=0; i<mapIterator->second.size(); i++){
+      HitReferences newHitPlaneMap = tmpHitPlaneMap;
+      newHitPlaneMap[mapIterator->first] = i;
+      PointInPlaneList newVector = tmpHitVector;
+      newVector.push_back(mapIterator->second.at(i));
```

Comment:
use [i] as you have already checked that i is in the range of the vector

Reply:
done